### ISSUE TITLE
Support for player name changes

### ISSUE THREAD

**Issue by SoniEx2** [minetest/minetest#12476]
## Problem

A clear and concise description of what the problem is.
ie: Why is this needed?
Ex. I'm always frustrated when [...]

Related: #12475 

There are... a lot of reasons why the ability to change player name independently of mod state is important. It's actually a bit impractical to list them all here.

With #12475 in particular, this would allow changing the player name from "singleplayer" to a more relevant name at runtime. But even without #12475, this remains relevant if a player decides to host a game from their singleplayer world, today, from the main menu - particularly because mods break.

## Solutions

There would need to be an API break and the introduction of player IDs.

## Alternatives

Deprecating "singleplayer" would solve hosting a world that's also a singleplayer world. But it's not very helpful long-term.

## Additional context

https://github.com/kdeldycke/awesome-falsehood/blob/7c4ef81641d59889de2171a61d373d7d8f2cc4c2/readme.md#human-identity

**Comment by Emojigit** [comment]
Hmm... That's nearly impossible. All mods store user-specific data by using the username as key. Therefore, it's a waste of time to change them or to code migration scripts, either manually or automatically.

**Comment by Zughy** [comment]
Yeah, basically every mod should stop storing players by name and store them by their UUID. It means breaking every possible mod around, like... ouch?

**Comment by appgurueu** [comment]
> There are... a lot of reasons why the ability to change player name independently of mod state is important. It's actually a bit impractical to list them all here.

That's what a *display name* is for. Mods *can already implement conventions for **display names***; in fact there have been multiple little fleshed out attempts by mods and games to establish displaying different names than the player names

> There would need to be an API break and the introduction of player IDs.

Wrong. Player names already server as player IDs and are superior in multiple ways (f.E., easier to remember). There's absolutely no need to deprecate them. Just implement *display names* on top *through mods*, which is 99% possible.

What would be possible if *display name support* was added on the engine side of things would be auto-generating the name (= ID) when joining servers (as poor mobile clients already do with their `PlayerXXX` accs).

> https://github.com/kdeldycke/awesome-falsehood/blob/7c4ef81641d59889de2171a61d373d7d8f2cc4c2/readme.md#human-identity

I don't see how that's relevant here. We are no hotel room reservation service. How to deal with *display names* is to be left up to mods either way - there are lots of nontrivial questions to be asked here: Which names should be allowed, how often may names be changed?

TL;DR: Leave this up to mods, continue using player names as IDs as to not break compat because there's absolutely zero need to break compat just to replace memorable names with some base64 or hex gibberish.

**Comment by SoniEx2** [comment]
Yeah we should shove UUIDs in the username field and add an actual username field.

**Comment by appgurueu** [comment]
> But also, no, you shouldn't get to store deadnames in your database.

![Clown](https://upload.wikimedia.org/wikipedia/commons/thumb/5/58/Clown_costume.jpg/220px-Clown_costume.jpg)

**Comment by SoniEx2** [comment]
It is, conveniently, also a GDPR violation. Names change, get over it.

**Comment by appgurueu** [comment]
> It is, conveniently, also a GDPR violation. Names change, get over it.

Implementing name changes does not help with the GDPR violation at all. We'd still need IDs which are identifiers just as well. You're completely missing the point and requesting ridiculous engine changes when there already is only a minimum of resources available. Make a PR or forget about it.

**Comment by SoniEx2** [comment]
IDs are identifiers but this isn't about identifiers. The GDPR actually has provisions for keeping information up-to-date - like names.

**Comment by rubenwardy** [comment]
I would like there to be support for changing usernames, but requiring all mods to break is a high price. There would need to be some compatibility, for example, older mods would still work but would consider name changes the same as a new user (as the name has changed).

A username is a username, and is always up to date. So I don't buy the GDPR argument.

A related issue would be the ability to use a full character set in names, as not everyone uses the roman alphabet. The reason this wasn't done is because of filename limitations (no longer an issue as we use sqlite) and also ease of moderation (could use a separate username or id instead)

**Comment by SoniEx2** [comment]
Sure, letting the existing API's usernames change, and adding an UID API, is an option. It'd cause strange behaviour with mods that aren't up to date with the new API, and auth would likely need to be reworked.

**Comment by appgurueu** [comment]
> A related issue would be the ability to use a full character set in names, as not everyone uses the roman alphabet. The reason this wasn't done is because of filename limitations (no longer an issue as we use sqlite) and also ease of moderation (could use a separate username or id instead)

Keep in mind that lifting any of these restrictions in usernames would be a breaking change as well due to the various messy ways mods use to work with usernames, *including* some mods still using a file-based storage (which is reasonable considering that current player meta refs are only available while the player is online).

**Comment by rubenwardy** [comment]
> That still makes it an API break, even if a minor one.

It's livable, and worst case the same as creating a new user. Although, one problem would be what happens if a new user logs in with the old username. You'd probably want to keep all previous usernames reserved for security reasons

> Keep in mind that lifting any of these restrictions in usernames would be a breaking change as well due to the various messy ways mods use to work with usernames, _including_ some mods still using a file-based storage (which is reasonable considering that current player meta refs are only available while the player is online).

That's a good point. It's especially a problem if they do manual text manipulation, as 1 character isn't one letter in UTF-8

**Comment by SoniEx2** [comment]
> You'd probably want to keep all previous usernames reserved for security reasons

Why not make it configurable? (Maybe with an expiry timer?)

**Comment by NathanSalapat** [comment]
Rather than allowing name changes, just tell the player that wants to change their name to log in with the new name and old at the same time and then manually give all their stuff to their new player, and transfer area ownerships. The only thing this method wouldn't handle is player data stored elsewhere. That sounds like a player issue though, they should have picked the name they wanted from the getgo.

**Comment by SoniEx2** [comment]
Well, they can't. When you make a singleplayer world your name is "singleplayer".

**Comment by NathanSalapat** [comment]
My bad, I thought you were talking about players changing their names on servers.

**Comment by SoniEx2** [comment]
That too. Singleplayer is just a server.

**Comment by NathanSalapat** [comment]
Yes singleplayer is a server, but what reason would you have to want to change your name in singleplayer. You almost never see it.
If you created a local world and wanted to make it into a server later you can set the name of the admin to anything you want. It's easy enough to grep through any file based mod storage to find instances of singleplayer and change them to whatever you want. If the mod is using a database it's probably a little more difficult, but I expect you can still change them. You'd only have to do this once.

**Comment by Arcitec** [comment]
In my opinion, the login username should be the account username. Then mods can still use the username as the unique ID.

And then the display name should be a separate value, changeable via /name in-game.

Have a new API for getDisplayName for mods that want to display the correct name.

**Comment by sfence** [comment]
> In my opinion, the login username should be the account username. Then mods can still use the username as the unique ID.
> 
> And then the display name should be a separate value, changeable via /name in-game.
> 
> Have a new API for getDisplayName for mods that want to display the correct name.

See `core.set_nametag_attributes` Lua API function.

`set_nametag_attributes({name = "YourViewName"})` -> this should be enough to change displayed name. No changes to the engine are needed. It can be done in Lua mod.

**Comment by Arcitec** [comment]
@sfence Oh really, that's very cool!
